x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSessionCCSUtils.java

```diff
@@ -217,34 +209,47 @@ static void updateExecutionInfoWithClustersWithNoMatchingIndices(EsqlExecutionIn
                 }
             } else {
                 // handles local cluster (when no concrete indices requested) and skip_unavailable=true clusters
-                EsqlExecutionInfo.Cluster.Status status;
-                ShardSearchFailure failure;
+                Cluster.Status status;
+                Exception failureException;
                 if (c.equals(RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY)) {
-                    status = EsqlExecutionInfo.Cluster.Status.SUCCESSFUL;
-                    failure = null;
+                    status = Cluster.Status.SUCCESSFUL;
+                    failureException = null;
                 } else {
-                    status = EsqlExecutionInfo.Cluster.Status.SKIPPED;
-                    failure = new ShardSearchFailure(new VerificationException("Unknown index [" + indexExpression + "]"));
+                    status = Cluster.Status.SKIPPED;
+                    failureException = new VerificationException("Unknown index [" + indexExpression + "]");
                 }
-                executionInfo.swapCluster(c, (k, v) -> {
-                    var builder = new EsqlExecutionInfo.Cluster.Builder(v).setStatus(status)
-                        .setTook(new TimeValue(0))
-                        .setTotalShards(0)
-                        .setSuccessfulShards(0)
-                        .setSkippedShards(0)
-                        .setFailedShards(0);
-                    if (failure != null) {
-                        builder.setFailures(List.of(failure));
-                    }
-                    return builder.build();
-                });
+                markClusterNoShards(executionInfo, c, status, failureException);
             }
         }
         if (fatalErrorMessage != null) {
             throw new VerificationException(fatalErrorMessage);
         }
     }
 
+    /**
+     * Mark cluster with a default cluster state with the given status and potentially failure from exception.
+     * Most metrics are set to 0 except for "took" which is set to the total time taken so far.
+     */
+    public static void markClusterNoShards(
```

Comment:
Can we add/change:

1. method name to `markClusterWithFinalStateAndNoShards`
2. add a comment that this should only be called with a final/terminal state (state != RUNNING)
3. add an assert that state != RUNNING

I don't want future users of this to lose the context. Since we show intermediate states now in the async-search response while the search is still running, we need to only set successful and failed shard counts at the end of the search for each remote (i.e., when setting the final state).